lib/Smime.php

```diff
@@ -123,13 +141,101 @@ public function addPersonalPublicKey($key, $signkey = false)
+
+        // check if a private key already exists
+        $check  = $prefs->getValue('smime_private_key');
+
+        // it there is a private key, these will be unset first and then the new one will be loaded
+        if (empty($check)) {
+            $GLOBALS['prefs']->setValue($prefName, $val);
+        } else {
+            $this->unsetSmimePersonal();
+            $GLOBALS['prefs']->setValue($prefName, $val);
+        }
+    }
+
+    /**
+     * Adds am extra personal public key to the extra keys table.
+     *
+     * @param string|array $key  The public key to add.
+     * @param int $private_key_id   PrivateKeyId to add the Public key to.
+     * previously: addExtraPersonalPublicKey
+     */
+    public function addExtraPublicKey($pref_name = 'smime_private_key', $public_key)
+    {
+        /* Get the user_name  */
+        // TODO: is there a way to only use prefs?
+        $user_name = $GLOBALS['registry']->getAuth();
+
+        /* Build the SQL query. */
+        $query = 'UPDATE imp_smime_extrakeys SET public_key=? WHERE private_key_id = ?'; //...
+        $values = [$public_key, $private_key_id, $pref_name];
+
+        try {
+            $this->_db->insert($query, $values);
+        } catch (Horde_Db_Exception $e) {
+            return $e;
+        }
+    }
+
+
+    /**
+     * Adds am extra personal private key to the extra keys table.
+     *
+     * @param string $pref_name To be removed... TODO.
+     * @param string|array $key  The private key to add.
+     * @param string|array $key  The public key to add (optional).
+     */
+    public function addExtraPersonalPrivateKey($pref_name = 'smime_private_key', $private_key)
+    {
+        /* Get the user_name  */
+        // TODO: is there a way to only use prefs?
+        $user_name = $GLOBALS['registry']->getAuth();
+
+        /* Build the SQL query. */
+        $query = 'INSERT INTO imp_smime_extrakeys (pref_name, user_name, private_key) VALUES (?, ?, ?)';
+        $values = [$pref_name, $user_name, $private_key];
+
+
+        try {
+            $this->_db->insert($query, $values);
+        } catch (Horde_Db_Exception $e) {
+            return $e;
+        }
+    }
+
+    /**
+     * Adds am extra personal keys to the extra keys table.
+     *
+     * @param string $pref_name To be removed... TODO.
+     * @param string|array $key  The private key to add.
+     * @param string|array $key  The public key to add.
+     */
+    public function addExtraPersonalKeys($private_key, $public_key, $password, $pref_name = 'smime_private_key')
+    {
+        /* Get the user_name  */
+        // TODO: is there a way to only use prefs?
+        $user_name = $GLOBALS['registry']->getAuth();
+
+        if (!empty($public_key) && !empty($private_key) && !empty($password)) {
+            /* Build the SQL query. */
+            $query = 'INSERT INTO imp_smime_extrakeys (pref_name, user_name, private_key, public_key, privatekey_passwd) VALUES (?, ?, ?, ?, ?)';
```

Comment:
Mmm not sure how pretty it is. Situation is like this:
Currently `unsetSmimePersonal` is the only method that callls `addExtraPersonalKeys` and it checks if the same key is already set. Because: `unsetSmimePersonal` is the only function that can currenlty add extra keys, currenlty one cannot upload extra keys directly.... which would probably be better

Reply:
ok I added a button to set keys into the extra table directly. Added the check to `addExtraPersonalKeys` accordingly